packages/button/src/button.tsx

```diff
@@ -118,17 +117,12 @@ export const Button = forwardRef<ButtonProps, "button">((props, ref) => {
     ...(!!group && { _focus }),
   }
 
-  const [isButton, setIsButton] = React.useState(!as)
-  const refCallback = React.useCallback((node: HTMLElement | null) => {
-    if (node?.tagName !== "BUTTON") {
-      setIsButton(false)
-    }
-  }, [])
+  const isButton = Boolean(!as)
```

Comment:
This solution is a bit too simple. The `as` prop could be a React Component which renders a button.
Checking the `node.tagName` is necessary.

The optional chaining operator `node?.tagName` returns `undefined | string`. If it is `undefined` we do not want to take any action, because we cannot be sure if it is not a button.

The previous logic with this change should to the trick:

```js
if (node && node.tagName !== "BUTTON") {
  setIsButton(false)
}
```

Reply:
Can we use OR operator to check the case `as="button"`? On the pervious logic, if `as` exists `isButton = false` and there's no trigger to set it to true.

Something like: 
```tsx
const isButton = Boolean(!as) || Boolean(as === "button")
```

Reply:
Good catch, the initialState would be `false` if an `as` prop would be provided.

```js
const [isButton, setIsButton] = React.useState(!as || as === 'button')

// ...

if (node && node.tagName !== "BUTTON") { 
  setIsButton(false) 
}
```

What do you think?


I am getting curious: @segunadebayo do you have any insights why the `tagName` check was introduced?

Reply:
I've implemented and extend test cases, i'm also curious about tagName check seems unnecessary to me

Reply:
I've updated the code to allow the user-defined `type` attribute to win. The reason we added a tagName check is to ensure we don't have `type="button"` to an element that isn't a button.

Technically, it is the work of the user to pass the correct type but since we're setting an internal default type of "button", it's up to us to ensure we don't add that to a non-button element.

Typical example: `<Button as="a">` should not have our internal `type=button` default attribute.